<a href="https://colab.research.google.com/github/myumbla3/AI_Assignment03/blob/main/Yumbla_A3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#import section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn import svm
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import BayesianRidge
from sklearn import linear_model

# Part 1. Loading the dataset

In [ ]:
#Using pandas load the dataset (load remotely, not locally)
data=pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
# Output the first 15 rows of the data
data.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [ ]:
# Display a summary of the table information (number of datapoints, etc.)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


# Part 2. Spliting the dataset

In [ ]:
# Take the pandas dataset and split it into our features (X) and label (y)
features = data[["Temperature °C","Mols KCL"]]
label = data ["Size nm^3"]

In [ ]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(features, label, test_size = 0.1, random_state = 0)
print(f'Holding out {len(X_test)} records for testing. ')
print(f'Using {len(X_train)} records for training.')

Holding out 100 records for testing. 
Using 900 records for training.


# Part 3. Perform a Linear Regression

In [ ]:
# Use sklearn to train a model on the training set
from sklearn.linear_model import LinearRegression
from sklearn.datasets import make_regression


lin_reg = LinearRegression()
lin_reg.fit(X_train,y_train)


prediction=lin_reg.predict(X_test)

print(f"Prediction on entire test data: {prediction}")

Prediction on entire test data: [ 134891.57109752  566848.1770532  1241137.60975794  -24367.1696409
  -70657.3340248   384956.44708746  467392.47066236  960873.69862144
    8407.39619923   51611.35989487  838231.36519831 1232720.46674085
  901563.3931183  -118835.20674597  188176.20244545 1158467.7057509
  665263.26735435  560560.92590561  163940.53886357  469055.93661764
 1419180.64745835  -84401.29716064  555956.53417346 1401489.68980634
  450621.8221688  1161403.05700777   45784.96014151  408948.47482721
  421868.97874553  645969.69402768  467893.44392697  208517.41869172
 1438187.73449241  708330.20055877 1400047.62029314  711132.70221441
  942380.06884206 1116186.8384737  1257396.19462681  189369.12027644
   71374.08771584  870102.22808014   94087.37936383  407174.28131093
  271473.01984826  388791.32041263  237873.60116061  790013.37060675
  112125.67358907  512139.47602221  905252.99153208 -163237.6627926
 1056768.59266956  757865.71816238  422587.22624213  838109.54727724
 1168

In [ ]:
# Create a sample datapoint and predict the output of that sample with the trained model

predictions=lin_reg.predict(np.array([[0,1]]))
predictions

array([-399299.78591432])

In [ ]:
#Report on the score for that model, in your own words (markdown, not code) explain what the score means
print(f"Score: {lin_reg.score(X_test,y_test)}")

Score: 0.8761646752736478


This score is showing the accuracy of our model, so our model has 87.6% accuracy.



In [ ]:
## Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
lin_reg.coef_,lin_reg.intercept_

(array([ 863.58108791, 1006.12741921]), -400305.91333353304)

$$y = {863.58109x_1 + 1006.12742x_2 -400305.91333353304}$$

# Part 4. Use Cross Validation

In [ ]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

scores = cross_val_score(lin_reg, X_test, y_test, cv=5)
scores

array([0.85851856, 0.85005287, 0.82953903, 0.80537035, 0.86740215])

Report on their finding and their significance:
The significance of the score computed at all 5 CV iteration is the score method of the estimator in this case lin_reg.

# Part 5. Using Polynomial Regression

In [ ]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(2)
#Report on the metrics and output the resultant equation as you did in Part 3.
X_train=poly.fit_transform(X_train)
X_test=poly.fit_transform(X_test)
model=BayesianRidge()
model.fit(X_train,y_train)
print(f"Score: {model.score(X_train,y_train)}")
print(f"Coefficient, intercept: {model.coef_,lin_reg.intercept_}")

Score: 1.0
Coefficient, intercept: (array([ 0.00000000e+00,  1.20000000e+01, -1.12373527e-07, -9.54969437e-12,
        2.00000000e+00,  2.85714287e-02]), -400305.91333353304)


The score is now 1 which shows the accuracy has improved drastically since that means 100% accuracy.

$$y=12a-1.12374e-07b-9.54969e-12a^2+2.00000ab+.0285714b^2$$